src/Build/BackEnd/Components/SdkResolution/OutOfProcNodeSdkResolverService.cs

```diff
@@ -63,21 +64,27 @@ public override void PacketReceived(int node, INodePacket packet)
         /// <inheritdoc cref="ISdkResolverService.ResolveSdk"/>
         public override SdkResult ResolveSdk(int submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, string solutionPath, string projectPath, bool interactive, bool isRunningInVisualStudio)
         {
+            bool wasResultCached = false;
+
+            MSBuildEventSource.Log.OutOfProcSdkResolverServiceRequestSdkPathFromMainNodeStart(submissionId, sdk.Name, solutionPath, projectPath);
+
             // Get a cached response if possible, otherwise send the request
-            var sdkResult = _responseCache.GetOrAdd(
+            Lazy<SdkResult> sdkResultLazy = _responseCache.GetOrAdd(
                 sdk.Name,
-                key =>
-                {
-                    var result = RequestSdkPathFromMainNode(submissionId, sdk, loggingContext, sdkReferenceLocation, solutionPath, projectPath, interactive, isRunningInVisualStudio);
-                    return result;
-                });
+                key => new Lazy<SdkResult>(() => RequestSdkPathFromMainNode(submissionId, sdk, loggingContext, sdkReferenceLocation, solutionPath, projectPath, interactive, isRunningInVisualStudio)));
```

Comment:
nit: Why not initialize `wasResultCached` with `true` and then reset to `false` when the factory callback runs? The `Lazy` is not super expensive but could be easily avoided.

```C#
            var sdkResultLazy = _responseCache.GetOrAdd(
                sdk.Name,
                key =>
                {
                    wasResultCached = false;
                    return RequestSdkPathFromMainNode(submissionId, sdk, loggingContext, sdkReferenceLocation, solutionPath, projectPath, interactive, isRunningInVisualStudio);
                }
```

Reply:
(There is a race and `wasResultCached` could be falsely negative with certain thread scheduling. Both with `Lazy` and the change I'm suggesting. I'm assuming it is OK.)

Reply:
Oh, I see, it's actually desired to not run `RequestSdkPathFromMainNode` multiple times for the same node. I take my suggestion back then. The race could be addressed by moving `wasResultCached = false` to the lambda passed to `Lazy`.

Reply:
Good point, threads are not blocked until they access the `Value` property so multiple threads could see `false` when in fact only one actually did the work.  That actually brings up an interesting point, do we want to differentiate calls that had to wait in order to get a cached result?  I'd have to have two booleans, one that said it wasn't cached before and one that says it ended up waiting for a cached result.  What are your thoughts?

Reply:
Yeah that's true, out-of-proc nodes do not evaluate projects in parallel as far as I know.  I'll leave it as-is for now